### PR TITLE
PRT-859-Provider-self-test-check-CORS

### PR DESCRIPTION
**Pull request by candostyavuz**

### CODE REVIEW

File: protocol/rpcprovider/cors_test.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	os.Exit(code)
+}
+
+func TestPerformCORSCheckFail(t *testing.T) {
+	endpoint := epochstoragetypes.Endpoint{
+		IPPORT: "localhost:8080", // pointing to the simple server
+	}
+
+	err := performCORSCheck(endpoint)
+	if err == nil {
```
Comment (ranlavanet) [inline protocol/rpcprovider/cors_test.go:35]: usually in tests we check for the exact error with require

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/rpcprovider/cors_test.go (hunk 1 above)
Comment (candostyavuz) [reply protocol/rpcprovider/cors_test.go]: fixed: 7d088c4c230787aed17e7b57bf0a78ad6e8ed8a8

File: protocol/rpcprovider/testing.go [hunk 2]
```diff
@@ -41,6 +42,43 @@ func validatePortNumber(ipPort string) string {
 	return ipPort
 }
 
+func performCORSCheck(endpoint epochstoragetypes.Endpoint) error {
+	utils.LavaFormatDebug("Checking CORS", utils.Attribute{Key: "endpoint", Value: endpoint})
+	// Construct the URL for the RPC endpoint
+	endpointURL := "http://" + endpoint.IPPORT // Assuming HTTP,
```
Comment (ranlavanet) [inline protocol/rpcprovider/testing.go:48]: https not http. the endpoints needs to have a certificate.

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/rpcprovider/testing.go (hunk 2 above)
Comment (ranlavanet) [inline protocol/rpcprovider/testing.go:48]:
```suggestion
	endpointURL := "https://" + endpoint.IPPORT // Providers must have HTTPS support,
```

File: protocol/rpcprovider/testing.go [hunk 3]
```diff
@@ -41,6 +42,43 @@ func validatePortNumber(ipPort string) string {
...
+
+	// Send an HTTP OPTIONS request to the endpoint
+	req, err := http.NewRequest("OPTIONS", endpointURL, nil)
```
Comment (ranlavanet) [inline protocol/rpcprovider/testing.go:51]: can we also validate GET / POST / PUT as well?

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/rpcprovider/testing.go (hunk 2 above)
Comment (ranlavanet) [reply protocol/rpcprovider/testing.go:48]: also fix the test afterwards as the s will make it fail.

Comment (candostyavuz) [reply protocol/rpcprovider/testing.go]: fixed here: 67f120c8f0ff8ec4740e5253469ef5cdc2702806

Comment (candostyavuz) [reply protocol/rpcprovider/testing.go]: fixed here: 67f120c8f0ff8ec4740e5253469ef5cdc2702806

File: protocol/rpcprovider/testing.go (hunk 3 above)
Comment (candostyavuz) [reply protocol/rpcprovider/testing.go]: fixed here: 52b5248c745ed6f749843612d14167b4d43e8e77

**Review by ranlavanet — APPROVED** [review]

**Pull request merged**